Support skopeo auth when inspecting prev-ref

cthtrifork opened the pull request:
This solves https://github.com/hhd-dev/rechunk/issues/19

cthtrifork committed:
Support skopeo auth when inspecting prev-ref

antheas commented:
Mmmm. Can you try to pull the skopeo command outside? Then people can do podman login in their action runner and it will work. I am afraid this might cause regressions and it only applies to the github registry

cthtrifork commented:
> Mmmm. Can you try to pull the skopeo command outside? Then people can do podman login in their action runner and it will work. I am afraid this might cause regressions and it only applies to the github registry

Would it not be a big refactor as you are calling ` sudo podman run` multiple times in your `action.yaml` ?
This is the reason i added it inside the `3_chunk.sh` script but also added environment variable checks to not break existing calls.

I agreed it is limited to the github registry and I could add custom inputs to solve this.

The easiest way to solve your requirement would be probably inject the auth file instead - would that be preferable or am I missing out on something

antheas commented:
The skopeo command just downloads the manifest file. If you pull it outside the podman script it will be able to access the creds of your action and will work. You can also opt to add an action parameter that allows you to specify a path for the manifest and if it exists skip the skopeo call

cthtrifork committed:
supply a registry_auth_file for chunking

cthtrifork commented:
I tried implementing the registry_auth_file, but I can change it again to volume mount a previous manifest file which has been pulled outside your action. The volume mount path would then be a input action

cthtrifork committed:
vscode did not save

antheas commented:
Actually the way you did it might be good enough. Does it work?

cthtrifork committed:
ensure default path

cthtrifork commented:
Just tested it. It works!

antheas commented:
Does it work if the auth file does not exist?

cthtrifork committed:
REGISTRY_AUTH_FILE_ARGS

cthtrifork committed:
moved \

cthtrifork committed:
broken syntax

cthtrifork commented:
I think you are right. I will test more scenarios. I forgot that volume mounts only create nonexisting directories but not nonexisting files

cthtrifork committed:
simplify REGISTRY_AUTH_FILE_ARGS

cthtrifork committed:
$

cthtrifork committed:
fix quoting

cthtrifork committed:
switch back to bash array

cthtrifork commented:
@antheas  bash was being unfriendly but I have tested multiple scenaries including happy path, non existing auth file etc.

cthtrifork commented:
Let me know if you prefer a rebase or you are just going to squash

cthtrifork commented:
@antheas What do you prefer? or do you want a different approach altogether. Right now I am using my fork but I would like to use upstream.

miku-orb commented:
If it really is needing the auth to access the registry just to download the manifest, without any actual contents later on, then perhaps indeed specifying just the path to a manifest file (and have that take precedence over trying to pull it based on `prev-ref`) could be perhaps the best, with some mention in the docs.

But in fact, turns out, that doesn't even seem to need any changes, because you'll notice these lines in `3_chunk.sh`:
```sh
PREV_NAME=${PREV_NAME:=previous}
PREV_MANIFEST=${PREV_MANIFEST:=./${PREV_NAME}.manifest.json}
```

Also notice `WORKDIR /workspace` in `Dockerfile` and `-v "${{ github.workspace }}:/workspace"` when running the action.

So turns out all that I need to do is to run this before running my action (assuming the 3 env vars I'm using are populated by me before and I'm in default workspace dir):
```sh
# Work around Rechunker not being able to use private registry auth details
skopeo inspect "docker://${{ env.IMAGE_REGISTRY }}/${{ env.IMAGE_NAME }}:${{ env.DEFAULT_TAG }}" > previous.manifest.json
```

Buuut, annoyingly:
```sh
if [ -n "$PREV_REF" ]; then
    echo "PREV_REF is set, downloading manifest"
    for i in $(seq 1 5); do
        skopeo inspect docker://${PREV_REF} > $PREV_MANIFEST && break
        echo "Failed to download previous manifest, retrying in 3 seconds"
        sleep 3
    done
```
So I need to also unset `prev-ref` completely, since it takes precedence without checking if the manifest file already exists. And not only will that waste time on retries, but will also wipe the target file due to the redirect even if it failed otherwise. It doesn't seem like `prev-ref` is used for anything else, so that should be ok.

But yeah, just tested it like that and it works with current upstream rechunk.

TL;DR put the manifest JSON into your work directory as `previous.manifest.json`, remove `prev-ref` parameter and then it just works!

cthtrifork commented:
@miku-orb  cool thanks for getting back to this. It works as you say and I will use your upstream together with `previous.manifest.json` and a removed `prev-ref`

antheas commented:
Sorry for not merging this earlier. Yes prev-ref is only used to pull the manifest. It is a valid alternative to pull the manifest as I suggested earlier and it seems the code already allows for that

miku-orb commented:
> cool thanks for getting back to this. It works as you say and I will use your upstream 

@cthtrifork Hey, it sounds like you confused me with the original author with the wording. I'm just some random guy who happened to have the same issue (meant to report it a few months before but forgot), then saw the PR wasn't merged so decided to tinker around to look for some workaround, and found one decent enough lol
